tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/CategoricalCrossentropy.java

```diff
@@ -0,0 +1,104 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.framework.metrics;
+
+import org.tensorflow.Operand;
+import org.tensorflow.framework.losses.Losses;
+import org.tensorflow.framework.metrics.impl.LossInterface;
+import org.tensorflow.framework.metrics.impl.MeanMetricWrapper;
+import org.tensorflow.op.Ops;
+import org.tensorflow.types.family.TNumber;
+
+/**
+ * A Metric that computes the categorical cross-entropy loss between true labels and predicted
+ * labels.
+ *
+ * <p>This is the crossentropy metric class to be used when there are multiple label classes (2 or
+ * more). The labels should be given as a one_hot representation. eg., When labels values are <code>
+ * [2, 0, 1]</code>, the labels Operand contains = <code>[[0, 0, 1], [1, 0, 0], [0, 1, 0]]
+ * </code>.
+ *
+ * @param <U> the data type for the predictions.
+ * @param <T> The data type for the metric result
+ */
+public class CategoricalCrossentropy<U extends TNumber, T extends TNumber>
+    extends MeanMetricWrapper<U, T> implements LossInterface<T> {
+
+  private final boolean fromLogits;
+  private final float labelSmoothing;
+  private final int axis;
+
+  /**
+   * Creates a CategoricalCrossentropy metric that Computes the crossentropy metric between the
+   * labels and predictions.
+   *
+   * <p>Uses a {@link Losses#CHANNELS_LAST} for the channel axis.
```

Comment:
`CHANNELS_FIRST` and `CHANNELS_LAST` do apply to images, and yes, `CategoricalCrossentropy`,  can be applied to other things than images, that is why the overloaded constructors are provided, where the axis defaults to `CHANNELS_LAST`.

Reply:
Although the term "channel" comes from vision work, I have noticed it bleeding over into NLP papers when teams explore combining convolutions and Transformer-based architectures. See, for example, [ConvBERT: Improving BERT with Span-based Dynamic Convolution](https://arxiv.org/abs/2008.02496) and [Depthwise Separable Convolutions for Neural Machine Translation](https://arxiv.org/abs/1706.03059). So although I think it's nicer to avoid the term for functionality that isn't squarely aimed at vision, I'd propose not overly worrying about non-vision folks encountering it from time to time.

Reply:
Ok, all this said, I finally spent some time with google and then with the Python source trying to understand how our use of the term "channel" for categorical cross-entropy is anchored into other work and other code. I do see the uses in the Python tensorflow code that I believe correspond to @JimClarke5 's use of "channel". Notably python/keras/backend.py:4893:
~~~ python
@keras_export('keras.backend.categorical_crossentropy')
@dispatch.add_dispatch_support
@doc_controls.do_not_generate_docs
def categorical_crossentropy(target, output, from_logits=False, axis=-1):
  """Categorical crossentropy between an output tensor and a target tensor.

  Args:
      target: A tensor of the same shape as `output`.
      output: A tensor resulting from a softmax
          (unless `from_logits` is True, in which
          case `output` is expected to be the logits).
      from_logits: Boolean, whether `output` is the
          result of a softmax, or is a tensor of logits.
      axis: Int specifying the channels axis. `axis=-1` corresponds to data
          format `channels_last`, and `axis=1` corresponds to data format
          `channels_first`.
~~~

But as I google around and read more code, I find myself believing that the use of "channel" here comes from very specialized vision work -- basically image segmentation.

Most of Python's keras categorical-crossentropy APIs don't even support a choice of axis. The one I found that does take an axis parameter (`tf.keras.losses.sparse_categorical_crossentropy`) documents it as "the dimension along which the entropy is computed." And indeed, we adopt this terminology in our own implementation of `sparseCategoricalCrossentropy` at org/tensorflow/framework/losses/Losses.java:506. Should we switch to this terminology everywhere?